packages/storage/amplify_storage_s3_dart/lib/src/storage_s3_service/transfer/database/database_html.dart

```diff
@@ -0,0 +1,72 @@
+import 'dart:html';
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:amplify_storage_s3_dart/src/storage_s3_service/transfer/database/database_stub.dart'
+    as stub;
+import 'package:amplify_storage_s3_dart/src/storage_s3_service/transfer/database/transfer_data.dart';
+import 'package:meta/meta.dart';
+
+@visibleForTesting
+
+/// a key prefix used by [TransferDatabase] when adding a key value pair into local storage.
+/// it is an identifier to differentiate the data stored by [TransferDatabase] into local storage from others.
+const localStorageKeyPrefix = 'amplify_storage_transfer_data_';
+
+///{@macro amplify_storage_s3_dart.transfer_database}
+class TransferDatabase implements stub.TransferDatabase {
+  ///{@macroamplify_storage_s3_dart.transfer_database_token}
+  static const token = Token<TransferDatabase>();
+
+  @protected
+  @visibleForTesting
+  @visibleForOverriding
+
+  /// the browser window local storage.
+  final localStorage = window.localStorage;
+
+  @override
+  Future<int> deleteTransferRecords(String uploadId) async {
+    var result = 0;
+    final keysToDelete = <String>[];
+    localStorage.entries
+        .where((element) => element.key.startsWith(localStorageKeyPrefix))
+        .forEach((element) {
+      final transferData = TransferData.fromJsonString(element.value);
+      if (transferData.uploadId == uploadId) {
+        keysToDelete.add(element.key);
+      }
+    });
+    for (final key in keysToDelete) {
+      if (localStorage.containsKey(key)) {
+        localStorage.remove(key);
+        result++;
+      }
+    }
+    return result;
+  }
+
+  @override
+  Future<List<TransferData>> getMultipartUploadRecordsCreatedBefore(
+    DateTime dateTime,
+  ) async {
+    final result = <TransferData>[];
+    localStorage.entries
+        .where((element) => element.key.startsWith(localStorageKeyPrefix))
+        .forEach((element) {
+      final value = element.value;
+      final transferData = TransferData.fromJsonString(value);
+      if (DateTime.parse(transferData.createdAt).isBefore(dateTime)) {
+        result.add(transferData);
+      }
+    });
+    return result;
+  }
+
+  @override
+  Future<String> insertTransferRecord(TransferData data) async {
+    final key = localStorageKeyPrefix + uuid();
```

Comment:
If you use the uploadId as the key you can make some performance improvements in `delete`.

Reply:
uploadId may not be unique so using a prefix and uuid is to make sure the key is unique.